plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -15,116 +15,115 @@
 
 import com.google.common.collect.ImmutableList;
 import io.airlift.slice.Slice;
-import io.airlift.slice.SliceInput;
 import io.trino.spi.exchange.ExchangeSource;
 
 import javax.annotation.Nullable;
-import javax.annotation.concurrent.GuardedBy;
-import javax.crypto.SecretKey;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
-import java.net.URI;
-import java.util.Iterator;
 import java.util.List;
-import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
 
-import static com.google.common.base.Preconditions.checkArgument;
-import static java.lang.String.format;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static io.airlift.concurrent.MoreFutures.whenAnyComplete;
+import static java.lang.Math.min;
 import static java.util.Objects.requireNonNull;
 
 public class FileSystemExchangeSource
         implements ExchangeSource
 {
-    private final FileSystemExchangeStorage exchangeStorage;
-    @GuardedBy("this")
-    private final Iterator<URI> files;
-    @GuardedBy("this")
-    private final Iterator<Optional<SecretKey>> secretKeys;
-
-    @GuardedBy("this")
-    private SliceInput sliceInput;
-    @GuardedBy("this")
-    private boolean closed;
-
-    public FileSystemExchangeSource(FileSystemExchangeStorage exchangeStorage, List<URI> files, List<Optional<SecretKey>> secretKeys)
+    private final List<ExchangeStorageReader> readers;
+    private volatile boolean closed;
+
+    public FileSystemExchangeSource(
+            FileSystemExchangeStorage exchangeStorage,
+            List<ExchangeSourceFile> sourceFiles,
+            int maxPageStorageSize,
+            int exchangeSourceConcurrentReaders)
     {
-        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
-        checkArgument(requireNonNull(files, "files is null").size() == requireNonNull(secretKeys, "secretKeys is null").size(),
-                format("number of files (%d) doesn't match number of secretKeys (%d)", files.size(), secretKeys.size()));
-        this.files = ImmutableList.copyOf(files).iterator();
-        this.secretKeys = ImmutableList.copyOf(secretKeys).iterator();
+        requireNonNull(exchangeStorage, "exchangeStorage is null");
+        int numReaders = min(sourceFiles.size(), exchangeSourceConcurrentReaders);
+        ConcurrentLinkedQueue<ExchangeSourceFile> sourceFileQueue = new ConcurrentLinkedQueue<>(sourceFiles);
```

Comment:
nit: With `ArrayBlockingQueue` the advantage is that there's no need to create a wrapper object for every element but it uses locking internally. Though `ConcurrentLinkedQueue` is lock free. It doesn't feel like accesses to this queue will be highly contended, I wonder if it is better to go with `ArrayBlockingQueue`, and if we see any issues with lock contention we can replace it with an array based lock free implementation.